bdai_ros2_wrappers/bdai_ros2_wrappers/filters.py

```diff
@@ -215,15 +222,19 @@ def __init__(self, upstream: Filter, fn: Callable) -> None:
 
         Args:
             upstream: the upstream message filter.
-            fn: adapter implementation as a callable.
+            fn: a callable that takes messages as arguments and returns some
+            data to be signaled (i.e. propagated down the filter chain).
+            If none is returned, no message signaling will occur.
         """
         super().__init__()
         self.fn = fn
         self.connection = upstream.registerCallback(self.add)
 
     def add(self, *messages: Any) -> None:
         """Adds new `messages` to the adapter."""
-        self.signalMessage(self.fn(*messages))
+        result = self.fn(*messages)
```

Comment:
@amessing-bdai ended up going with the trivial implementation. There is a way, an `async` way for `fn` to nicely embed complex logic in a functional way, but it was more trouble than it was worth. If one needs lots of bookkeeping and control flow for an adapter, it may be about time to write a new filter.